### PR TITLE
Remove AnnData/Seurat file FF and migrate misc files to new type (SCP-4567, SCP-4732)

### PR DESCRIPTION
**Pull request by ehanna4**
TLDR: Remove AnnData/Seurat file upload FF and migrate misc files to their new appropriate type

- Misc file types like `Other` and `Documentation` will no longer accept `'.Rds', '.rds'` or `'.h5', '.h5ad'`

- `AnnData` and `Seurat` file upload sections in the wizard will not be feature flagged anymore and will be GA

- Any file uploaded in the categories` 'Other', 'Documentation',` or ` 'Analysis Output'` that end in `h5ad` or `h5` are migrated to type `AnnData` and those ending in `rds` or `Rdata` will migrate to type `'Seurat'`

To Test:
1. BEFORE pulling this branch
2. Get together an `.rds` and a `.h5ad` files to test with, I used this study in HCA which has an `.rds` and a `.h5ad` file to download https://data.humancellatlas.org/explore/projects/83f5188e-3bf7-4956-9544-cea4f8997756/get-curl-command
3. Start up your local server `./rails_local_setup.rb && source config/secrets/.source_env.bash && rails s`
4. Open up localhost in a browser and sign in
5. Navigate to a study you own and go to the upload wizard
6. Start the rails job uploads by opening another terminal and running `./rails_local_setup.rb && source config/secrets/.source_env.bash && bin/rails jobs:work`
7. Upload the `.rds` file and a `.h5ad` files to the `misc` file section choosing either 'Other' or 'Documentation' for the file type. 
8. Once that has completed open up another terminal window and start up a rails console with `rails c` 
9. In the console run `StudyFile.where(upload_file_name: /\.(h5ad)$/).pluck(:file_type)` This will print out the file types of any files ending in `.h5ad` and run `StudyFile.where(upload_file_name: /\.(rds)$/).pluck(:file_type)` to get any `.rds` ending files. You should see 'Other' or 'Documentation' at this point
Screen Shot 2022-10-07 at 11 16 23 AM

1. NOW Pull this branch
2. Kill the local server you've got running
3. Open up a new terminal window  and run the DB migration `bin/rails db:migrate`
5. Start up your local server
8. Open up localhost and sign in
10. Navigate to the same study you just uploaded to and go to the wizard
11. Observe that you should see now the `AnnData` and `Seurat` sections in the wizard
12. In the rails console rerun these commands `StudyFile.where(upload_file_name: /\.(h5ad)$/).pluck(:file_type)` and run `StudyFile.where(upload_file_name: /\.(rds)$/).pluck(:file_type)` and observe that the filetypes are now `Seurat` and `AnnData` as appropriate
Screen Shot 2022-10-07 at 11 22 58 AM

----------------------------------------------------------------
Interesting note is any invalid files that were in misc section did not migrate
Screen Shot 2022-10-06 at 5 13 58 PM

### CODE REVIEW

File: app/javascript/components/upload/UploadWizard.jsx [hunk 1]
```diff
@@ -69,14 +69,13 @@ export function RawUploadWizard({ studyAccession, name }) {
...
-  const allowAnnDataAndSeuratFileUploads = serverState?.feature_flags?.upload_seurat_and_anndata
 
-  if (allowAnnDataAndSeuratFileUploads && !STEPS.includes(AnnDataStep)) {
+  if (!STEPS.includes(AnnDataStep)) {
     STEPS.splice(8, 0, AnnDataStep)
     NON_VISUALIZABLE_STEPS.splice(0, 0, AnnDataStep)
   }
 
-  if (allowAnnDataAndSeuratFileUploads && !STEPS.includes(SeuratStep)) {
+  if ( !STEPS.includes(SeuratStep)) {
```
Comment (eweitz) [inline app/javascript/components/upload/UploadWizard.jsx:78]:
```suggestion
  if (!STEPS.includes(SeuratStep)) {
```

File: app/javascript/components/upload/upload-utils.js [hunk 2]
```diff
@@ -249,8 +249,8 @@ const plainTextExtensions = ['.txt', '.tsv', '.text', '.csv']
 const mtxExtensions = ['.mtx', '.mm', '.txt', '.text']
 const imageExtensions = ['.jpeg', '.jpg', '.png', '.bmp']
 const miscExtensions = ['.txt', '.text', '.tsv', '.csv', '.jpg', '.jpeg', '.png', '.pdf',
-  '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx', '.zip', '.loom', '.h5', '.h5ad', '.h5an',
-  '.ipynb', '.Rda', '.rda', '.Rds', '.rds']
+  '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx', '.zip', '.loom', '.h5an',
```
Comment (eweitz) [inline app/javascript/components/upload/upload-utils.js:252]: I think we can safely remove "h5an", as I can find no relevant reference to this file extension.

Its existence in this list derives from [a request on 2020-03-10 in our #scp Slack channel](https://broadinstitute.slack.com/archives/C9YSEMSFR/p1583863287031200), and searching Slack for "h5an" returns [one prior hit in #scp-implementation from 2019-03-15](https://broadinstitute.slack.com/archives/CBEHTH601/p1552673056000300) with a link to https://scanpy.readthedocs.io/en/latest/api/scanpy.read_10x_h5.html.  That link no longer mentions that extension, and the Scanpy repo doesn't mention it, so at most I suspect it was an equivalent alternative to .h5 or .h5ad that briefly existed and is now no longer used at all.

```suggestion
  '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx', '.zip', '.loom',
```

File: app/javascript/components/upload/upload-utils.js [hunk 3]
```diff
@@ -249,8 +249,8 @@ const plainTextExtensions = ['.txt', '.tsv', '.text', '.csv']
...
+  '.ipynb', '.Rda', '.rda']
```
Comment (eweitz) [inline app/javascript/components/upload/upload-utils.js:253]: Blocker: .Rda and .rda were both added for Seurat support, simultaneously with .Rds and .rds.  For consistency, we should remove the former here since we're removing the latter.

Context:
* Original user request: https://broadinstitute.zendesk.com/agent/tickets/143271
* Related PR: https://github.com/broadinstitute/single_cell_portal_core/pull/1002

**Review by eweitz — APPROVED** [review]
Code looks generally good!

I note a trivial functional blocker, based on a review of original user requests and past related work.  I also note a similar issue, and a nit.

File: db/migrate/20221006132730_update_misc_file_type_for_anndata_and_seurat.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+  class UpdateMiscFileTypeForAnndataAndSeurat < Mongoid::Migration
+    def self.up
+      # For AnnData
+      StudyFile.where(:file_type.in => ['Other', 'Documentation', 'Analysis Output'], upload_file_name: /\.(h5ad|h5)$/).update_all(file_type: 'AnnData')
+      
+      # For Seurat
+      StudyFile.where(:file_type.in => ['Documentation', 'Other', 'Analysis Output'], upload_file_name: /\.(rds|Rdata)$/).update_all(file_type: 'Seurat')
```
Comment (eweitz) [inline db/migrate/20221006132730_update_misc_file_type_for_anndata_and_seurat.rb:7]: Blocker: see related comment from other review regarding .rda / .Rda.  Also, let's account for camel-case file extensions, which seem often idiomatic in the R community.

```suggestion
      StudyFile.where(:file_type.in => ['Documentation', 'Other', 'Analysis Output'], upload_file_name: /\.(rda|Rda|rds|Rda|Rdata)$/).update_all(file_type: 'Seurat')
```

File: app/javascript/components/upload/upload-utils.js [hunk 5]
```diff
@@ -249,8 +249,8 @@ const plainTextExtensions = ['.txt', '.tsv', '.text', '.csv']
...
 const sequenceExtensions = ['.fq', '.fastq', '.fq.tar.gz', '.fastq.tar.gz', '.fq.gz', '.fastq.gz', '.bam']
 const baiExtensions = ['.bai']
 const annDataExtensions = ['.h5', '.h5ad', 'h5ad1', '.hdf5']
```
Comment (eweitz) [inline app/javascript/components/upload/upload-utils.js:256]: ".h5ad1" seems like a typo.  Searching it in Slack, Zendesk, and Google returns no relevant results.

```suggestion
const annDataExtensions = ['.h5', '.h5ad', '.hdf5']
```

**Review by eweitz — COMMENTED** [review]
Another review pass revealed another trivial blocker, and a seeming functional typo from prior related work.

**Review by eweitz — CHANGES_REQUESTED** [review]
Following the [trivial blocker](https://github.com/broadinstitute/single_cell_portal_core/pull/1635#discussion_r992341143) I noted in the first review, I see we'll also want to add ".Rda" and ".rda" to `seuratExtensions` in `upload-utils.js`.

Since there are now 3 small functional blockers across as many reviews (sorry for the noise!) and involving code that I can't directly suggest changes on, I'll go ahead and mark this as "Request changes" to better ensure they're tracked.

File: db/migrate/20221006124530_retire_anndata_and_seurat_feature_flag.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+  class RetireAnndataAndSeuratFeatureFlag < Mongoid::Migration
+    def self.up
+      FeatureFlag.retire_feature_flag('upload_seurat_and_anndata')
+      FeatureFlag.retire_feature_flag('upload_seurat_and_anndata')
```
Comment (bistline) [inline db/migrate/20221006124530_retire_anndata_and_seurat_feature_flag.rb:4]: Blocker: this is likely a typo, but you've already retired the feature flag, so calling this twice throws an error (see below, Jean also encountered this error).  Also, in the `down` section, you're creating two instances of the `FeatureFlag` which would also throw an error.
```
==  RetireAnndataAndSeuratFeatureFlag: migrating ==============================
rails aborted!
StandardError: An error has occurred, 20221006124530 and all later migrations canceled:

FeatureFlag: 'upload_seurat_and_anndata' does not exist

    raise NameError, "FeatureFlag: '#{name}' does not exist" if feature_flag.blank?
    ^^^^^
/Users/bistline/Documents/Rails/single_cell_portal_core/app/models/feature_flag.rb:32:in `retire_feature_flag'
/Users/bistline/Documents/Rails/single_cell_portal_core/db/migrate/20221006124530_retire_anndata_and_seurat_feature_flag.rb:4:in `up'

Caused by:
NameError: FeatureFlag: 'upload_seurat_and_anndata' does not exist

    raise NameError, "FeatureFlag: '#{name}' does not exist" if feature_flag.blank?
    ^^^^^
/Users/bistline/Documents/Rails/single_cell_portal_core/app/models/feature_flag.rb:32:in `retire_feature_flag'
/Users/bistline/Documents/Rails/single_cell_portal_core/db/migrate/20221006124530_retire_anndata_and_seurat_feature_flag.rb:4:in `up'
Tasks: TOP => db:migrate
(See full trace by running task with --trace)

```

**Review by bistline — APPROVED** [review]
I'm going to proactively approve because the blocker I noted is simple to address (just removing duplicate lines) and the other concerns noted have already requested changes.  Once addressed this should be fine to merge.

File: app/javascript/components/upload/upload-utils.js (hunk 3 above)
Comment (ehanna4) [reply app/javascript/components/upload/upload-utils.js:253]: I presume then `.rda` and `.Rda` ought to be added to the Seurat extension list?
Also thank you for the context super helpful 👍

Comment (ehanna4) [reply app/javascript/components/upload/upload-utils.js:253]: oops just saw your other comment got it will add to the Seurat list
